cmd/network/setup_linux.go

```diff
@@ -0,0 +1,193 @@
+/*
+Copyright © 2023 SUSE LLC
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"os/exec"
+	"strconv"
+
+	"github.com/linuxkit/virtsock/pkg/vsock"
+	"github.com/sirupsen/logrus"
+	"github.com/vishvananda/netns"
+
+	"github.com/rancher-sandbox/rancher-desktop-networking/pkg/log"
+	rdvsock "github.com/rancher-sandbox/rancher-desktop-networking/pkg/vsock"
+)
+
+var (
+	debug           bool
+	vmSwitchPath    string
+	unshareArg      string
+	logFile         string
+	vmSwitchLogFile string
+)
+
+const (
+	nsenter            = "/usr/bin/nsenter"
+	unshare            = "/usr/bin/unshare"
+	vsockHandshakePort = 6669
+	vsockDialPort      = 6656
+)
+
+func main() {
+	flag.BoolVar(&debug, "debug", false, "enable additional debugging")
+	flag.StringVar(&vmSwitchPath, "vm-switch-path", "", "the path to the vm-switch binary that will run in a new namespace")
+	flag.StringVar(&vmSwitchLogFile, "vm-switch-logfile", "", "path to the logfile for vm-switch process")
+	flag.StringVar(&unshareArg, "unshare-arg", "", "the command argument to pass to the unshare program")
+	flag.StringVar(&logFile, "logfile", "/var/log/network-setup.log", "path to the logfile for network setup process")
+	flag.Parse()
+
+	if err := log.SetOutputFile(logFile, logrus.StandardLogger()); err != nil {
+		logrus.Fatalf("setting logger's output file failed: %v", err)
+	}
+
+	if debug {
+		logrus.SetLevel(logrus.DebugLevel)
+	}
+
+	if vmSwitchPath == "" {
+		logrus.Fatal("path to the vm-switch process must be provided")
+	}
+
+	if unshareArg == "" {
+		logrus.Fatal("unshare program arg must be provided")
+	}
+
+	// listenForHandshake blocks until a successful handshake is estabilished.
+	listenForHandshake()
+
+	logrus.Debugf("attempting to connect to the host on CID: %v and Port: %d", vsock.CIDHost, vsockDialPort)
+	vsockConn, err := vsock.Dial(vsock.CIDHost, vsockDialPort)
+	if err != nil {
+		logrus.Fatal(err)
+	}
+	logrus.Debugf("successful connection to host on CID: %v and Port: %d: connection: %+v", vsock.CIDHost, vsockDialPort, vsockConn)
+
+	// setup network namespace
+	ns, err := configureNamespace()
+	if err != nil {
+		logrus.Fatal(err)
+	}
+
+	if err := unshareCmd(ns, unshareArg); err != nil {
```

Comment:
As described offline, there's a race here: unshare can start running before the network namespace setup is complete (it already exists, but there's no `eth0` in it, never mind DHCP); however, this looks like it's irrelevant in practice, as we'll need to disable host-resolver when we're using this new network stack.

As a future improvement, it may be worthwhile to make vm-switch spawn unshare instead (after DHCP).